Fixes remaining issues with DynamicDependency attribute and adds tests

vitek-karas opened the pull request:
Ports over all DynamicDependencyAttribute tests we have in linker. All tests are now passing, two had to be partially disabled due to:
* copyused testing which NativeAOT doesn't really support
* known missing support for XML attribute files in AOT

Product changes:
* DynamicDependencyAttribute will produce a warning if it can't resolve the target of the dependency.
* Support for DynamicDependencyAttribute on fields

Small improvements to assembly checker - mainly ability to handle delegate types.

vitek-karas committed:
Fixes remaining issues with DynamicDependency attribute and adds tests

Ports over all DynamicDependencyAttribute tests we have in linker. All tests are now passing, two had to be partially disabled due to:
* copyused testing which NativeAOT doesn't really support
* known missing support for XML attribute files in AOT

Product changes:
* DynamicDependencyAttribute will produce a warning if it can't resolve the target of the dependency.
* Support for DynamicDependencyAttribute on fields

Small improvements to assembly checker - mainly ability to handle delegate types.

sbomer reviewed (approved):
LGTM, thank you!

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs:
```diff
@@ -26,135 +30,187 @@ public static void AddDependenciesDueToDynamicDependencyAttribute(ref Dependency
-                            // _context.LogWarning ($"Unresolved type '{typeName}' in DynamicDependencyAttribute", 2036, context);
-                            continue;
-                        }
+                        (factory.MetadataManager as UsageBasedMetadataManager)?.Logger?.LogWarning(
```
Is the `is` needed? Basically if we run this with other than the UsageBased manager, we're running this logic twice (first time scanner, second time real compilation). We should ideally not do that - it's useless work the second time.

```suggestion
                        ((UsageBasedMetadataManager)factory.MetadataManager).Logger.LogWarning(
```

vitek-karas replied:
Changed

MichalStrehovsky reviewed (commented):
I think there is one more issue remaining - will this keep XXX for the below? I think we need to also call into this algorithm from `FieldMetadataNode`.

```
typeof(Foo).GetFields();

class Foo
{
    [DynamicDependency(XXX)]
    public int Field;
}
```

MichalStrehovsky commented:
> I think there is one more issue remaining - will this keep XXX for the below? I think we need to also call into this algorithm from `FieldMetadataNode`.
> 
> ```
> typeof(Foo).GetFields();
> 
> class Foo
> {
>     [DynamicDependency(XXX)]
>     public int Field;
> }
> ```

And thinking about it more, also from `MethodMetadataNode` to cover:

```csharp
typeof(Foo).GetMethods();

class Foo
{
    [DynamicDependency(XXX)]
    public abstract void M();
}
```

marek-safar commented on src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DynamicDependencies/Dependencies/AssemblyDependency.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace Mono.Linker.Tests.Cases.DynamicDependencies.Dependencies
```
Do we want to duplicate all these tests? Wouldn't be better to have that in some shared location?

marek-safar replied:
I think if you put the files in the right location it'd be enough and save us from deleting them all in few weeks/months

vitek-karas replied:
I think the tricky bit is having them in the right location (don't think we agreed on one yet), and I don't want to have that discussion as part of a random unrelated PR.

marek-safar commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/DynamicDependencyAttributeAlgorithm.cs:
```diff
@@ -26,135 +30,187 @@ public static void AddDependenciesDueToDynamicDependencyAttribute(ref Dependency
+
+            if (fixedArgs.Length > 0 && fixedArgs[0].Value is string sigFromAttribute)
+            {
+                if (fixedArgs.Length == 1)
```
nit: use switch

vitek-karas commented:
> And thinking about it more, also from MethodMetadataNode

Technically that should not be needed since reflection enabled method is automatically callable and thus will have an entry point - which is what the current logic is tied to. (I added a simple test case similar to your example and it already works).
I'll look into this a little bit more though.

MichalStrehovsky commented:
> Technically that should not be needed since reflection enabled method is automatically callable and thus will have an entry point

Abstract methods don't have an entrypoint, but it's possible we're already taking the right codepaths. Maybe a test would still be useful if we don't already have such test.

vitek-karas committed:
PR feedback

Add tests for reflection-only accessed fields and methods and DynamicDependency on them. Fixed the problem with those tests by also hooking up FieldMetadataNote and MethodMetadataNode.

vitek-karas commented:
All the failures are known.
